addons/survey/controllers/main.py

```diff
@@ -118,7 +116,7 @@ def _get_access_data(self, survey_token, answer_token, ensure_token=True, check_
                 has_survey_access = True
             can_answer = bool(answer_sudo)
             if not can_answer:
-                can_answer = survey_sudo.access_mode == 'public'
+                can_answer = survey_sudo.access_mode == 'public' or ensure_token == 'survey_only'
```

Comment:
Don't we need this too?

```suggestion
                can_answer = survey_sudo.access_mode == 'public' or (
                    has_survey_access and ensure_token == 'survey_only')
```